src/transformers/generation_beam_constraints.py

```diff
@@ -0,0 +1,372 @@
+from abc import ABC, abstractmethod
+from typing import List, Optional, Union
+
+import torch
+
+from .file_utils import add_start_docstrings
+
+
+TOKEN_CONSTRAINT_DOCSTRING = r"""
+    [`Constraint`] enforcing that a specific token is included in the output.
+
+    Args:
+        token_ids (`torch.LongTensor`):
+            The sequence of tokens that must be generated by the output.
+"""
+
+PHRASAL_CONSTRAINT_DOCSTRING = r"""
+    [`Constraint`] enforcing that an ordered sequence of tokens is included in the output.
+
+    Args:
+        token_id (`int`):
+            The id of the token that must be generated by the output.
+"""
+
+
+class Constraint(ABC):
+    r"""Abstract base class for all constraints that can be applied during generation.
+    It must define how the constraint can be satisfied.
+
+    All classes that inherit Constraint must follow the requirement that
+
+    ```
+    completed = False
+    while(not completed):
+        _, completed = constraint.update(constraint.advance())
+    ```
+
+    will always terminate (halt).
+    """
+
+    def __init__(self):
+        # test for the above condition
+        self.test()
+
+    def test(self):
+        """
+        Tests whether this constraint has been properly defined.
+        """
+        counter = 0
+        completed = False
+        while not completed:
+            if counter == 1:
+                self.reset()
+            advance = self.advance()
+            assert self.does_advance(advance)
```

Comment:
We don't add new `assert` in the code base, especially not without an error message. Please put a test an raise the appropriate exception here.